File: java/Refactor2/src/main/java/no/bouvet/solid/srpdip/domain/orderoperations/CancelOrderOperation.java
package no.bouvet.solid.srpdip.domain.orderoperations;

import no.bouvet.solid.srpdip.Factory;
import no.bouvet.solid.srpdip.InventoryRepository;
import no.bouvet.solid.srpdip.OrderRepository;
import no.bouvet.solid.srpdip.domain.InventoryItem;
import no.bouvet.solid.srpdip.domain.Order;
import no.bouvet.solid.srpdip.domain.OrderItem;
import no.bouvet.solid.srpdip.domain.OrderItemState;
import no.bouvet.solid.srpdip.domain.OrderState;
import no.bouvet.solid.srpdip.messageinterface.RequestMessage;
import no.bouvet.solid.srpdip.messageinterface.ResponseMessage;
import no.bouvet.solid.srpdip.messageinterface.ResponseMessageFactory;

public class CancelOrderOperation  implements OrderOperation{

	public static Factory<CancelOrderOperation> factory = new Factory<>(CancelOrderOperation.class);
	
	private InventoryRepository inventoryRepository = InventoryRepository.factory.getInstance();
	private OrderRepository orderRepository = OrderRepository.factory.getInstance();

	private ResponseMessageFactory responseMessageFactory = new ResponseMessageFactory();
	
	@Override
	public ResponseMessage execute(RequestMessage request) {
		try {
			Order orderToCancel = orderRepository.orders.get(request.getOrderId());

			switch (orderToCancel.getState())
			{
			case SHIPPED:
				// msg cannot cancel shipped order
				return responseMessageFactory.createCancellationResponseMessage(request, orderToCancel,
						"Cannot cancel an order that has been shipped.");
			case CANCELLED:
				return responseMessageFactory.createCancellationResponseMessage(request, orderToCancel,
						"Order has already been cancelled.");
			default:
				for (OrderItem item : orderToCancel.getOrderItems())
				{
					InventoryItem inventoryItem = inventoryRepository.inventory.get(item.getItemCode());

					if (item.getState() == OrderItemState.FILLED)
						inventoryItem.setQuantityOnHand(
								inventoryItem.getQuantityOnHand() + item.getQuantity());

					item.setState(OrderItemState.CANCELLED);
				}

				orderToCancel.setState(OrderState.CANCELLED);

				// save inventory
				inventoryRepository.updateInventory();

				// save order
				orderRepository.updateOrders();

				return responseMessageFactory.createCancellationResponseMessage(request, orderToCancel,
						"Order has been cancelled and reserved inventory has been released.");
			}
		} catch (Exception ex) {
			// Trace.WriteLine("Exception during operation CancelOrder: " + ex,
			// "CancelOrderError");
			return responseMessageFactory.createErrorResponseMessage(request.getRequestId(), ex);
		}
	}

}

File: cs/Refactor1/SRP_DI_Workshop/Domain/OrderOperations/CancelOrderOperation.cs
using System;
using SRP_DI_Workshop.ServiceInterface;

namespace SRP_DI_Workshop.Domain.OrderOperations
{
    public class CancelOrderOperation : IOrderOperation
    {
        public InventoryRepository InventoryRepository;
        public OrderRepository OrderRepository;
        private readonly LoggerService _loggerService = new LoggerService();
        private readonly ResponseMessageFactory _responseMessageFactory = new ResponseMessageFactory();

        public const Operation Id = Operation.Cancel_Order;

        public ResponseMessage ExecuteOperation(RequestMessage request)
        {
            try
            {
                Order orderToCancel = OrderRepository.GetOrder(request.OrderId);

                switch (orderToCancel.State)
                {
                    case OrderState.Shipped:
                        // msg cannot cancel shipped order
                        return _responseMessageFactory.CreateCancellationResponseMessage(request, orderToCancel,
                            "Cannot cancel an order that has been shipped.");
                    case OrderState.Cancelled:
                        return _responseMessageFactory.CreateCancellationResponseMessage(request, orderToCancel,
                            "Order has already been cancelled.");
                    default:
                        foreach (OrderItem item in orderToCancel.OrderItems)
                        {
                            InventoryItem inventoryItem = InventoryRepository.GetInventoryItem(item.ItemCode);

                            if (item.State == OrderItemState.Filled)
                                inventoryItem.QuantityOnHand += item.Quantity;

                            item.State = OrderItemState.Cancelled;
                        }

                        orderToCancel.State = OrderState.Cancelled;

                        // save inventory
                        InventoryRepository.UpdateInventory();

                        // save order
                        OrderRepository.UpdateOrders();

                        return _responseMessageFactory.CreateCancellationResponseMessage(request, orderToCancel,
                            "Order has been cancelled and reserved inventory has been released.");
                }
            }
            catch (Exception ex)
            {
                _loggerService.WriteLine("Exception during operation CancelOrder: " + ex, "CancelOrderError");
                return _responseMessageFactory.CreateErrorResponseMessage(ex);
            }
        }
    }
}